pkg/backend/httpstate/client/api.go

```diff
@@ -107,7 +112,32 @@ func pulumiAPICall(ctx context.Context, d diag.Sink, cloudAPI, method, path stri
 	path = cleanPath(path)
 
 	url := fmt.Sprintf("%s%s", cloudAPI, path)
-	req, err := http.NewRequest(method, url, bytes.NewReader(body))
+	var bodyReader io.Reader
+	if opts.GzipCompress {
+		// If we're being asked to compress the payload, go ahead and do it here to an intermediate buffer.
+		//
+		// If this becomes a performance bottleneck, we may want to consider marshaling json directly to this
+		// gzip.Writer instead of marshaling to a byte array and compressing it to another buffer.
+		logging.V(7).Infoln("compressing payload using gzip")
+		var buf bytes.Buffer
+		writer := gzip.NewWriter(&buf)
+		defer contract.IgnoreClose(writer)
+		if _, err := writer.Write(body); err != nil {
+			return "", nil, errors.Wrapf(err, "compressing payload")
+		}
+
+		// gzip.Writer will not actually write anything unless it is flushed.
+		if err := writer.Flush(); err != nil {
+			return "", nil, errors.Wrapf(err, "flushing compressed payload")
+		}
+
+		logging.V(7).Infof("gzip compression ratio: %f", float64(len(body))/float64(len(buf.Bytes())))
```

Comment:
Nit: Let's include the original payload size for context. As XX% compression doesn't mean much if it was only 6-bytes to begin with.

Reply:
done.